### PR TITLE
WebXRManager: Added depth sensing support

### PR DESCRIPTION
**Pull request by cabanier**
First pass at adding support for WebXR GPU based depth sensing.
Also updated the XR samples to request depth.

We can tweak shaders and parameters later.

*This contribution is funded by [Meta](https://meta.com)*

### CODE REVIEW

File: src/renderers/webgl/WebGLProgram.js [hunk 1]
```diff
@@ -510,6 +510,8 @@ function WebGLProgram( renderer, cacheKey, parameters, bindingStates ) {
 			parameters.useFog && parameters.fog ? '#define USE_FOG' : '',
 			parameters.useFog && parameters.fogExp2 ? '#define FOG_EXP2' : '',
 
+			parameters.occlusion ? '#define USE_OCCLUSION' : '#define USE_OCCLUSION',
+
```
Comment (CodyJasonBennett) [inline src/renderers/webgl/WebGLProgram.js:514]: Both sides are equal; the right-hand side should be `''`.

Comment (cabanier) [reply src/renderers/webgl/WebGLProgram.js:514]: I know :-)
That is why it's WIP and still a draft PR

**Comment by hybridherbst** [comment]
Super cool!

Out of curiosity, have you done a perf/quality comparison between:
- passing depth into each material as in this PR
- drawing a fullscreen pass that renders depth only and sets z buffer values so others test against  that
- doing that z writing from e.g. the background material (so it stays one pass)

I’d think writing the z buffer directly loses the fuzziness but should be faster (can use early z discard, no extra texture fetch per fragment, etc)

**Comment by cabanier** [comment]
> Super cool!
> 
> Out of curiosity, have you done a perf/quality comparison between:
> 
> * passing depth into each material as in this PR
> * drawing a fullscreen pass that renders depth only and sets z buffer values so others test against  that, for example right from the background material
> 
> I’d think writing the z buffer directly loses the fuzziness but should be faster (can use early z discard, no extra texture fetch per fragment, etc)

I have not done a perf comparison. This way seems to integrate nicely with three, although it's a bit special since it's a renderer feature.
We can revisit more optimal approaches later. I can also investigate how this is done in unity and unreal.

**Comment by mrdoob** [comment]
@hybridherbst

> drawing a fullscreen pass that renders depth only and sets z buffer values so others test against that, for example right from the background material

I feel like this approach is much more straight forward and easier to maintain, regardless of the performance implications.

I think I would try implementing this approach first before modifying all the shaders for all materials.

File: src/renderers/webxr/WebXRManager.js [hunk 2]
```diff
@@ -730,6 +739,32 @@ class WebXRManager extends EventDispatcher {
...
+					if (glDepthTexture === null) {
+						glDepthTexture = depthData.texture;
+
+						const depthTexture = new Texture();
+						const texProps = renderer.properties.get(depthTexture);
+						texProps.__webglTexture = glDepthTexture;
+
+						if (( depthData.depthNear != session.renderState.depthNear ) || ( depthData.depthFar != session.renderState.depthFar )) {
+
+							session.updateRenderState( {depthNear: depthData.depthNear, depthFar: ( depthData.depthFar > 10000 ? 10000 : depthData.depthFar ) } );
```
Comment (cabanier) [inline src/renderers/webxr/WebXRManager.js:754]: this seems like a WebXR bug

Comment (toji) [reply src/renderers/webxr/WebXRManager.js:754]: Spec looks like it allows infinity:

> If activeState’s [depthFar](https://immersive-web.github.io/webxr/#dom-xrrenderstate-depthfar) is greater than session’s [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane) set activeState’s [depthFar](https://immersive-web.github.io/webxr/#dom-xrrenderstate-depthfar) to session’s [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane).

And

> The [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane) SHOULD be greater than 1000.0 (and MAY be infinite).

So if it's not working then it's a Chromium implementation issue. Idea is to allow `depthFar: Infinity`, right?

Comment (cabanier) [reply src/renderers/webxr/WebXRManager.js:754]: > So if it's not working then it's a Chromium implementation issue. Idea is to allow `depthFar: Infinity`, right?

Yes. If I pass infinity, I get an error saying the value is non-finite

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -1600,7 +1615,7 @@ class WebGLRenderer {
 
 			const lightsStateVersion = lights.state.version;
 
-			const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, object );
+			const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, object, _occlusion );
```
Comment (cabanier) [inline src/renderers/WebGLRenderer.js:1618]: @mrdoob is there a better way than passing in a renderer variable?

File: src/renderers/webgl/WebGLProgram.js (hunk 1 above)
Comment (cabanier) [reply src/renderers/webgl/WebGLProgram.js]: @CodyJasonBennett fixed!

**Comment by cabanier** [comment]
@mrdoob @Mugen87 would you mind taking a look?

File: src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+export default /* glsl */`
+#ifdef USE_OCCLUSION
+    if (gl_FragColor.a > 0.0) {
+        int arrayIndex = 0;
+        vec2 depthUv;
+        if (gl_FragCoord.x>=depthWidth) {
```
Comment (snowzurfer) [inline src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:6]: is this checking which eye to read the texture for?

**Comment by snowzurfer** [comment]
fwiw it looks like on Unity, at least for custom shaders, [they suggest modifying the shaders and  calculating the occlusion in the fragment shader](https://github.com/oculus-samples/Unity-DepthAPI/tree/main#step-4-calculate-occlusions-in-fragment-shader):
```
half4 frag(v2f i) {
   UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(i);

   // this is something your shader will return without occlusions
   half4 fragmentShaderResult = someColor;

   // calculate UV for the depth texture lookup for occlusions
   float2 uv = i.positionNDC.xy / i.positionNDC.w;

   // pass UV and the current depth of the texel
   float occlusionValue = CalculateEnvironmentDepthOcclusion(uv, i.vertex.z);

   // consider early rejection to not write to depth if it's an opaque shader
   if (occlusionValue < 0.01) {
       discard;
   }

   // premultiply color and alpha by occlusion value
   // when it's 1 - color is not affected - virtual covers real
   // when it's 0 - texel is invisible - virtual is under real
   // when it's in between - texel is semi transparent
   fragmentShaderResult *= occlusionValue;

   return result;
}
```

File: src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js (hunk 4 above)
Comment (cabanier) [reply src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:6]: Yes. I considered passing it in as a value but that required deeper changes in the renderer.

**Comment by cabanier** [comment]
> fwiw it looks like on Unity, at least for custom shaders, [they suggest modifying the shaders and calculating the occlusion in the fragment shader](https://github.com/oculus-samples/Unity-DepthAPI/tree/main#step-4-calculate-occlusions-in-fragment-shader)

We can play with the shaders in follow up patches. It might even be interesting to vary which depth values are honored, or provide a color to overlay real world objects

**Comment by mrdoob** [comment]
@cabanier Is this the latest spec: https://www.w3.org/TR/webxr-depth-sensing-1/

Also, did you see [this feedback](https://github.com/mrdoob/three.js/pull/27154#issuecomment-1803354732
)?

**Comment by mrdoob** [comment]
> Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?

That's right. We render a quad with the depth data first and then OpenGL does the rest.
No need to modify any built-in shader.

> If so, that would destroy performance on headsets as it would introduce a depth resolve.

Do you mind elaborating this point?

**Comment by cabanier** [comment]
> > Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?
> 
> That's right. We render a quad with the depth data first and then OpenGL does the rest. No need to modify any built-in shader.

How would we use the quad with the depth data? Is there a way to do a depth compare between 2 textures? Note that the resolutions of the depth buffer and real world depth are different.

> > If so, that would destroy performance on headsets as it would introduce a depth resolve.
> 
> Do you mind elaborating this point?

Because we're rendering to a multisampled framebuffer, the depth buffer is very big and only needed while we're resolving the pixels. The resolved depth buffer is never used by the system so we make sure that it's invalidated at the very end of the render pass. That way it is not flushed to main memory which saves a lot of GPU time.

**Comment by cabanier** [comment]
> > Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?
> 
> That's right. We render a quad with the depth data first and then OpenGL does the rest. No need to modify any built-in shader.

@mrdoob if you can point me to how I could implement this, I will give it a shot.

**Comment by mrdoob** [comment]
On my Quest 3 with Chrome 118 this sample doesn't work (clicking in START AR does nothing):
https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html

Is there a flag I need to enable?

**Comment by cabanier** [comment]
> On my Quest 3 with Chrome 118 this sample doesn't work (clicking on START AR does nothing): https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html
> 
> Is there a flag I need to enable?

Are you on an [early access OS release](https://www.meta.com/en-gb/help/quest/articles/headsets-and-accessories/using-your-headset/quest-ptc/)? You need v60. 
If so, go to `chrome://flags`, enable `WebXR experiments` and restart the browser.

**Comment by cabanier** [comment]
> > On my Quest 3 with Chrome 118 this sample doesn't work (clicking on START AR does nothing): https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html
> > Is there a flag I need to enable?
> 

FYI even with the flag, that sample won't work because it uses the CPU version of depth sensing

**Comment by mrdoob** [comment]
Ah ah I see...

Unfortunately I do not have the bandwidth to figure out how the early access stuff works.

Lets hold this until v60 is out 👍

**Comment by cabanier** [comment]
> Lets hold this until v60 is out 👍

v60 is now out. Browser 30.3 is also out with experimental support. Go to `chrome://flags`, enable `WebXR experiments` and restart the browser.

I'm still unconvinced that a depth prepass is a practical solution.  See  https://keaukraine.medium.com/efficient-rendering-of-soft-particles-on-mobile-gpus-9beb856fcfbc or https://developer.arm.com/documentation/101897/0301/Optimizing-application-logic/Avoid-using-depth-prepasses on why that is

**Comment by KTRosenberg** [comment]
Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.

**Comment by cabanier** [comment]
> Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.

No, there are no `adb` flags; just the `WebXR experiments` in `chrome://flags`

**Comment by KTRosenberg** [comment]
> > Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.
> 
> No, there are no `adb` flags; just the `WebXR experiments` in `chrome://flags`

I might have misunderstood your comment above here: https://github.com/mrdoob/three.js/pull/27154#issuecomment-1831447301

This was just referring to the old example for some Android phones.

So the new example here should work on Quest 3?

**Comment by cabanier** [comment]
> This was just referring to the old example for some Android phones.
> So the new example here should work on Quest 3?

Yes, it's only for Quest 3. I *think* it should be easy to make it work for android but I don't have a device to test it on.

**Comment by cabanier** [comment]
@mrdoob any thoughts? Did you try this out?

**Comment by hybridherbst** [comment]
@cabanier a depth prepass usually means rendering the entire scene once to a depth target, then binding that for compares. The article correctly notes that this is slow on mobile devices. However, here, depth would be written to the depth buffer once (usually fast) and then normal rendering happens with regular ZTest (super fast). Still think it’s worth a try!

File: src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+            arrayIndex = 1;
+            depthUv = vec2((gl_FragCoord.x-depthWidth)/depthWidth, gl_FragCoord.y/depthHeight);
+        } else {
+            depthUv = vec2(gl_FragCoord.x/depthWidth, gl_FragCoord.y/depthHeight);
+        }
+        float assetDepthM = gl_FragCoord.z;
+        
+        //float occlusion = Depth_GetOcclusion(depthColor, depthUv, assetDepthM, arrayIndex);
```
Comment (snowzurfer) [inline src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:14]: I think this is possibly a leftover comment? maybe we could remove it

**Review by snowzurfer — COMMENTED** [review]
one small comment

also, I think we should use @cabanier 's current approach rather than the depth prepass one, because without it we can't have soft edges, which are quite fundamental to making the occlusion look good.

I think that most (virtually all?) AR systems have soft edges since it looks much better. Even though it introduces a bit more complexity (imo not that much), it's a worthwhile tradeoff to get soft edges.

also like @cabanier said, the two depth maps would be of different size

**Comment by cabanier** [comment]
> @cabanier a depth prepass usually means rendering the entire scene once to a depth target, then binding that for compares. The article correctly notes that this is slow on mobile devices. However, here, depth would be written to the depth buffer once (usually fast) and then normal rendering happens with regular ZTest (super fast). Still think it’s worth a try!

It's never fast to write depth :-)
We were careful to code the browser and three.js so it never resolves depth. It's only used during the tile computation and then discarded so it never leaves the cache.

Writing regular or multisampled depth data and then reading back from it introduces too much overhead. This is also the reason why occlusion is happening at the application level and not the system compositor one and why postprocessing isn't practical.

File: src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js (hunk 5 above)
Comment (cabanier) [reply src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:14]: I left it here for experimentation. I suspect that we want to tweak the shader but happy to remove it.

**Comment by cabanier** [comment]
@mrdoob @Mugen87 depth sensing is now on by default in the Quest browser. Can we merge this PR?

**Comment by hybridherbst** [comment]
@cabanier I was able to try this out (finally!) but still had to turn the option on on v60. Looks very jaggy/pixelated even on Quest 3 but for some experiences this might be acceptable. Is there already a newer version than v60 that has it enabled by default?

Also, some questions:
1. can the depth transition "softness" be controlled per object?
2. can the depth effect be turned off per object/per material? From the code it looks like this would just be `material.occlusion = false;`, right?

**Comment by cabanier** [comment]
> @cabanier I was able to try this out (finally!) but still had to turn the option on on v60. Looks very jaggy/pixelated even on Quest 3 but for some experiences this might be acceptable.

True. There are better shaders than the simple ones I added. I was hoping that we experiment with them after landing the fundamentals.

>  Is there already a newer version than v60 that has it enabled by default?

It should be enabled by default in 31.1. Is that not the version you have?
OS v60 has support for depth sensing.

> 1. can the depth transition "softness" be controlled per object?

Not at the moment
> 2. can the depth effect be turned off per object/per material? From the code it looks like this would just be `material.occlusion = false;`, right?

Yes, for some content it might be desired that it's not occluded. (ie hands or the controller)

**Comment by Mugen87** [comment]
Closing in favor of https://github.com/mrdoob/three.js/pull/27586.

**Comment by hybridherbst** [comment]
@Mugen87 I'm not sure if this issue should be closed, the topic of "how to allow the shader-based version in a sample" is still very much open. Many effects can't be achieved with the v2 implementation and vice versa, both have their place.

**Comment by Mugen87** [comment]
I'm not sure the project is willing to support both paths especially since this PR required in-depth changes in the material system. Unless @mrdoob does not decide different, `three.js` will only offer what has been implement in https://github.com/mrdoob/three.js/pull/27586. 

My hope is it will be easier to support changes like this with `WebGPURenderer`.

**Comment by hybridherbst** [comment]
I mean like this – v2 in core, v1 as sample:
https://github.com/mrdoob/three.js/pull/27586#issuecomment-1899317734

Of course @cabanier or @mrdoob can open separate PRs for a sample that uses custom shaders, I just think it gets lost when this PR here is closed.

**Comment by cabanier** [comment]
> I mean like this – v2 in core, v1 as sample: [#27586 (comment)](https://github.com/mrdoob/three.js/pull/27586#issuecomment-1899317734)
> 
> Of course @cabanier or @mrdoob can open separate PRs for a sample that uses custom shaders, I just think it may get lost when this PR here is closed.

I will make an attempt to extend the WebXRManager and write a test that uses a custom shader.

**Pull request closed without merging**